pkg/daemon/config_drift_monitor.go

```diff
@@ -0,0 +1,378 @@
+type unitConfigDriftErr error
+
+type ConfigDriftMonitor interface {
+	Start(ConfigDriftMonitorOpts) error
+	Done() <-chan struct{}
+	IsRunning() bool
+	Stop()
+}
+
+type ConfigDriftMonitorOpts struct {
+	// Called whenever a config drift is detected.
+	OnDrift func(error)
+	// The currently applied MachineConfig.
+	MachineConfig *mcfgv1.MachineConfig
+	// The Systemd dropin path location.
+	// Defaults to /etc/systemd/system
+	SystemdPath string
+	// Channel to report unknown errors
+	ErrChan chan<- error
+}
+
+// Holds the Config Drift Watcher and ensures we only have a single instance
+// running at a given time.
+type configDriftMonitor struct {
+	cdw    *configDriftWatcher
+	mu     sync.Mutex
+	stopCh chan struct{}
+}
+
+// Implements the Config Drift Watcher
+type configDriftWatcher struct {
+	ConfigDriftMonitorOpts
+	watcher   *fsnotify.Watcher
+	filePaths sets.String
+	wg        sync.WaitGroup
+	stopCh    chan struct{}
+}
+
+// Holds a single Config Drift Watcher and starts / stops it as necessary while
+// ensuring that only a single Config Drift Monitor is running at any given
+// time (within its scope).
+func NewConfigDriftMonitor() ConfigDriftMonitor {
+	return &configDriftMonitor{
+		mu:     sync.Mutex{},
+		stopCh: make(chan struct{}),
+	}
+}
+
+// Determines if a Config Drift Watcher is running; useful for avoiding certain
+// logic such as emitting a startup event.
+func (c *configDriftMonitor) IsRunning() bool {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.cdw != nil
+}
+
+// Starts the Config Drift Monitor if not already started. It will create a new
+// Config Drift Monitor instance, if necessary.
+func (c *configDriftMonitor) Start(opts ConfigDriftMonitorOpts) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if c.cdw != nil {
+		return nil
+	}
+
+	var err error
+
+	// We don't have a Config Drift Watcher running, create one.
+	c.cdw, err = newConfigDriftWatcher(opts)
+	if err != nil {
+		return err
+	}
+
+	c.cdw.start()
+
+	return nil
+}
+
+// Provides a channel for listening if the currently running Config Drift
+// Monitor has been stopped. Can be used similar to <-ctx.Done().
+func (c *configDriftMonitor) Done() <-chan struct{} {
+	out := make(chan struct{})
+
+	go func() {
+		select {
+		case <-c.stopCh:
+			out <- struct{}{}
+			close(out)
+			return
+		}
+	}()
+
+	return out
+}
+
+// Stops the currently running Config Drift Monitor and deletes its instance so
+// that it cannot be reused. Also sends a signal to the Done() listener.
+func (c *configDriftMonitor) Stop() {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if c.cdw != nil {
+		c.cdw.stop()
+		c.cdw = nil
+		c.stopCh <- struct{}{}
+	}
+}
+
+// Creates a new config drift watcher for a given machineconfig and registers
+// a callback for when a drift occurs.
+func newConfigDriftWatcher(opts ConfigDriftMonitorOpts) (*configDriftWatcher, error) {
+	if opts.OnDrift == nil {
+		return nil, fmt.Errorf("no ondrift function attached")
```

Comment:
partially feel like this phrasing "no ondrift function" isn't very clear

Reply:
I agree, I'll cogitate on this a bit.